### PR TITLE
[https://nvbugs/6631019][fix] Add a `torch_compiling(enable)` contextmanager to `_torch/utils.py` and lower…

### PR DESCRIPTION
**Pull request by trtllm-agent**
## Summary
- **Root cause**: `_capture_key` runs the encoder block loop from `load_weights` without scoping the process-global `is_torch_compiling` flag, so a stale `True` from a prior engine routes attention through `create_attn_outputs`, which reads `attention_metadata` from `extra_attrs` that nothing binds at capture time.
- **Fix**: Add a `torch_compiling(enable)` contextmanager to `_torch/utils.py` and lower the flag for the duration of `_capture_key`'s warmup+capture, alongside the `torch.inference_mode()` already there for the same reason, restoring the caller's value on exit; drop this bug's DGX_B200 waiver.
- **Original test**: `pytest tests/integration/defs/accuracy/test_llm_api_pytorch.py::TestLlama3_1_8BInstruct::test_nvfp4_kv[v2_kv_cache=False-attn_backend=TRTLLM-torch_compile=True] tests/integration/defs/accuracy/test_llm_api_pytorch_multimodal.py::TestNanoV3Omni::test_auto_dtype[fp8_mmmu_encoder_cuda_graph] -v`
- Automated fix generated by repair-bot

## Test plan
- [x] Verify fix on the same GPU type as the original failure
- [x] Check for regressions in related tests

## Links
- Bug: https://nvbugs/6631019

## Dev Engineer Review

- Added `torch_compiling(enable)` to scope and restore the process-global Torch compile flag.
- Disabled Torch compilation during multimodal encoder CUDA graph warmup and capture.
- Disabled Torch compilation during RADIO block execution, including CUDA-graph replay and eager fallback.
- Restored the previous compile state on normal and exceptional exits.
- Removed the obsolete DGX_B200 waiver for `TestNanoV3Omni::test_auto_dtype[fp8_mmmu_encoder_cuda_graph]`.
- No configuration errors or unintended test-list scope changes were identified.

## QA Engineer Review

- Updated multimodal encoder capture tests.
- Added `test_radio_run_blocks_lowers_torch_compiling_for_eager_fallback`.
- Added `test_torch_compiling_restores_flag_after_exception`.
- Tests verify disabled Torch compilation during capture and RADIO eager fallback.
- Tests verify restoration of the caller’s Torch compile state.
- The affected integration test is covered by the removed waiver entry: `TestNanoV3Omni::test_auto_dtype[fp8_mmmu_encoder_cuda_graph]`.
- Verdict: sufficient.

### CODE REVIEW

File: tensorrt_llm/_torch/models/multimodal_encoder_graph.py [hunk 1]
```diff
@@ -383,7 +383,13 @@ def _capture_key(
...
 
         graph = torch.cuda.CUDAGraph()
-        with torch.inference_mode():
+        # `torch_compiling(False)` for the same reason as `inference_mode`: ambient state the encoder
+        # region must not inherit from whoever called us. Capture can be driven from `load_weights`,
+        # outside any engine forward, so the process-global compile flag may still carry a previous
+        # engine's value. `encoder_fn` is handed its metadata explicitly, so its layers must take
+        # the eager path rather than the custom op, which resolves metadata from `extra_attrs` that
+        # only an engine forward binds.
+        with torch.inference_mode(), torch_compiling(False):
```
Comment (brnguyen2) [inline tensorrt_llm/_torch/models/multimodal_encoder_graph.py:392]: This closes the capture path, but the eager path has the same exposure and no assert to catch it.

`maybe_run` returns `None` on a bucket miss, and `RADIOVisionModel._run_blocks` (`modeling_radio.py:917`) then calls `_run_blocks_eager` with the vision metadata — during a real engine forward, where the flag is legitimately `True` and `extra_attrs` is bound to the *LM's* `attn_layers`. RADIO's vision `Attention` is constructed with `config=self.model_config` (`modeling_radio.py:498`), so `register_to_config` stays `True` and `forward_impl` takes the custom-op branch; `create_attn_outputs` then resolves `"0"`, `"1"`, … against the LM's layers and the vision blocks silently get the LM's `seq_lens`/`num_contexts` with vision's `head_dim`. Qwen2-VL hit exactly this and fixed it by unregistering vision attention from `extra_attrs` (`modeling_qwen2vl.py:1557`).

Either move the `torch_compiling(False)` scope up to `_run_blocks` so capture and eager fallback are covered by one rule, or unregister the RADIO vision attention layers the way Qwen2-VL does. The current placement leaves the fallback broken for any image shape that misses every configured bucket.

File: tensorrt_llm/_torch/utils.py [hunk 2]
```diff
@@ -102,6 +102,22 @@ def is_torch_compiling() -> bool:
...
+    The flag is a plain module global, not thread- or context-local, so it
+    outlives the engine that set it. Code running a model region outside an
+    engine forward must establish the value rather than inherit it.
+    """
+    prev_enable = is_torch_compiling()
+    set_torch_compiling(enable)
+    try:
+        yield
+    finally:
+        set_torch_compiling(prev_enable)
```
Comment (brnguyen2) [inline tensorrt_llm/_torch/utils.py:118]: The `finally` restore on the exception path is the part most likely to regress silently later (e.g. someone "simplifying" this to a plain set/reset pair), and nothing exercises it. A three-line unit test in `tests/unittest/_torch/` — enter `torch_compiling(True)`, raise inside, assert the flag came back — would pin it.

File: tests/unittest/_torch/modeling/test_multimodal_encoder_graph.py [hunk 3]
```diff
@@ -538,39 +540,64 @@ def test_replay_matches_eager_across_sizes(cuda_device, make_cuda_runner, real_t
...
-        input_specs={
-            "x": EncoderGraphTensorSpec(shape=(HIDDEN,), dtype=torch.float32, token_dim=0),
-        },
-        output_specs={"y": 0},
-        config=config,
+    return encoder_fn
+
+
+@pytestmark_cuda
+def test_capture_uses_inference_mode_when_grad_enabled(cuda_device, make_cuda_runner):
```
Comment (brnguyen2) [inline tests/unittest/_torch/modeling/test_multimodal_encoder_graph.py:562]: The old test set `warmup_steps=2` explicitly, which is what made `[False, False, False]` (2 warmups + 1 capture) self-explanatory. Going through `make_cuda_runner` drops that and the length now silently tracks the Pydantic default. Either thread `warmup_steps` through the factory and pass it here, or assert on the property being tested — `assert observed and not any(observed)` — so a default change doesn't turn into a puzzling failure in an unrelated test.

**Review by brnguyen2 — APPROVED** [review]
Approving — the comments below are optional touch-ups, not blockers.

Root cause holds up under reading: `is_torch_compiling_flag` is the one bare module global among its neighbours (`_model_extra_attrs` is `threading.local`, the breakable-graph marker a `ContextVar`), `model_engine.py:801` raises it at engine init and never lowers it, and the encoder capture driven from `load_weights` runs before that point — so a reused worker sees the previous engine's True and `extract_extra_attrs` fires. The waiver removal is backed by that mechanism rather than a coincidental green run, which is the right bar.

Two things to settle before merge:

- The fix covers only capture. The same vision blocks run eagerly at request time on a bucket miss (`modeling_radio.py:917`), where the flag is legitimately True — see the inline comment.
- There is another change in flight that lowers the same flag from the test-harness side (#17931, `[test] reset compile mode in reused MPI workers`). Both are worth keeping, but only one should remove the waiver line — coordinate so the second one doesn't land a conflicting `waives.txt` edit.

`torch_compiling` is internal, so no docs/changelog owed.

File: tensorrt_llm/_torch/models/multimodal_encoder_graph.py (hunk 1 above)
Comment (allisonlim-nv) [reply tensorrt_llm/_torch/models/multimodal_encoder_graph.py:392]: fixed

File: tests/unittest/_torch/modeling/test_multimodal_encoder_graph.py (hunk 3 above)
Comment (allisonlim-nv) [reply tests/unittest/_torch/modeling/test_multimodal_encoder_graph.py:562]: It was also applied to the adjacent test:
```
test_capture_lowers_torch_compiling_and_restores_it in tests/unittest/_torch/modeling/test_multimodal_encoder_graph.py:599.
```

File: tensorrt_llm/_torch/utils.py (hunk 2 above)
Comment (allisonlim-nv) [reply tensorrt_llm/_torch/utils.py:118]: fixed

File: tests/unittest/_torch/test_utils.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+# limitations under the License.
+
+import pytest
+
+from tensorrt_llm._torch.utils import is_torch_compiling, torch_compiling
+
+pytestmark = pytest.mark.cpu_only
+
+
+def test_torch_compiling_restores_flag_after_exception() -> None:
```
Comment (2ez4bz) [inline tests/unittest/_torch/test_utils.py:23]: Is this test file included in any existing test lists in CI?

**Review by 2ez4bz — APPROVED** [review]

File: tests/unittest/_torch/test_utils.py (hunk 4 above)
Comment (allisonlim-nv) [reply tests/unittest/_torch/test_utils.py:23]: No it wasn't, I just added unittest/_torch/test_utils.py to l0_cpu; it now maps to both CPU Generic stages.

**Review by cascade812 — APPROVED** [review]
